### PR TITLE
Allow packages without repository

### LINKED ISSUE #30: Include all packages, even if they don't have repos
The build process for this package currently omits packages that don't have a `repository` field.

What if we changed it to preserve all the packages, and just use `null` values for packages that don't have a repository? Here's an example of what that would look like:

```js
{
  foo: 'https://github.com/someuser/foo',
  bar: 'https://github.com/someuser/bar',
  baz: null
}
```

If we made this change, this package would then implicitly include all the public package names in the npm registry, and it would make the process for updating the [`all-the-package-names`](https://github.com/nice-registry/all-the-package-names) package way easier, without any duplicate npm registry code between the two projects. It would be a matter of updating the `all-the-package-names` build and release process to install `all-the-package-repos` then output `Object.keys(require('all-the-package-repos'))`.

@jimaek what do you think?

### PR DESCRIPTION
**Pull request by zarianec**
This is the follow-up of #30 and #31. This PR allows packages without a repository to be added to the `packages.json` with a `null` value.

resolve #30

### CODE REVIEW

File: scripts/update.js [hunk 1]
```diff
@@ -213,20 +213,17 @@ const apply = (change) => {
...
     stats.deletes += 1
     return delete packages[name]
   }
 
-  const url = extractUrl(change)
-
-  if (!url) {
-    stats.invalid += 1
-    return
-  }
```
Comment (eridal) [inline scripts/update.js:229]: It was nice to have that stat in the metadata, can we keep it? it would be useful for, in ie, updating the readme.

File: scripts/update.js [hunk 2]
```diff
@@ -328,6 +325,10 @@ const extractDomain = (repoUrl) => {
 }
 
 const updateRepoStats = (url, delta) => {
+  if (typeof url !== 'string') {
+    return
+  }
```
Comment (eridal) [inline scripts/update.js:330]: Are you sure this will keep the count correctly over time? A package that was `+1` for having a url but is now null should be `-1`.

File: scripts/update.js (hunk 1 above)
Comment (zarianec) [reply scripts/update.js:229]: Yeah, I was thinking about it. We can calculate "invalid" packages based on `packages.json` since we keep all of them now. But the idea to have it pre-calculated in metadata is probably better.

File: scripts/update.js (hunk 2 above)
Comment (zarianec) [reply scripts/update.js:330]: Yes, this is correct. The logic you described is handled in the `apply` function. This check is needed to simplify the code there and to avoid bugs if a "bad" URL was passed as an argument.
I would like to add the tests for it but it will require significant changes to the `update.js`. 
Also, @LucasScotta is working on tests for stats so probably I will wait for his changes to avoid conflicts.

File: scripts/update.js [hunk 3]
```diff
@@ -213,19 +213,20 @@ const apply = (change) => {
...
+    updateRepoStats(curr, -1)
 
     stats.deletes += 1
     return delete packages[name]
   }
 
-  const url = extractUrl(change)
+  // TODO: this must be simplified as soon as standard linter
+  //  will get updated since it doesn't support ?? operator
+  //  const url = extractUrl(change) ?? null
```
Comment (zeke) [inline scripts/update.js:224]: 👍🏼

File: test.js [hunk 4]
```diff
@@ -17,11 +17,11 @@ describe('repos', () => {
...
   })
 
-  it('is always a URL', function () {
-    this.timeout(10 * 1000)
+  it('is always a URL or null', function () {
+    this.timeout(30 * 1000)
     const urls = Object.values(repos)
     urls.forEach(url => {
-      expect(isUrl(url), `${url}`).to.eq(true)
+      expect(url === null || isUrl(url), `${url}`).to.eq(true)
```
Comment (zeke) [inline test.js:24]: What about making sure that both URLs and nulls exist? Something like this:

```js
expect(urls.some(url => isUrl(url)).to.eq(true)
expect(urls.some(url => url === null).to.eq(true)
```

**Review by zeke — APPROVED** [review]
Left a minor refactoring suggestion, but this looks good to me. I will defer to @eridal for a final approval of this.

File: test.js (hunk 4 above)
Comment (zarianec) [reply test.js]: good point, added it to test. Thanks.

**Comment by zarianec** [comment]
Updated cache to store the packages without repository as well.

**Comment by zeke** [comment]
This looks good to me. What do you think, @eridal?

**Comment by eridal** [comment]
I have a few concerns with this proposal:

**The metadata is not handled correctly**, and over time the collected stats will deviate from the packages.

Currently the sum of repos in the metadata is exactly the number of packages:

```js
const packages = require('./data/packages.json')
const metadata = require('./data/metadata.json')

Object.keys(packages).length === Object.values(metadata.repos).reduce((a, b) => a + b, 0)
```

There are a few changes required to get those numbers correctly:

1. in [`apply`](https://github.com/nice-registry/all-the-package-repos/blob/master/scripts/update.js#L211), when a change gets processed we need to update correctly the counter (`+1`/`-1`) depending if it is an insert/update/delete. My suggestion is to make  [`extractUrl`](https://github.com/nice-registry/all-the-package-repos/blob/master/scripts/update.js#L229) simply return the string url, so that later we can parse/validate the url and separate invalid/nulls urls.

1. in [`updateRepoStats`](https://github.com/nice-registry/all-the-package-repos/blob/master/scripts/update.js#L335) we want to allow the counter to have impact when the url is null. My suggestion is to create a new entry in the [`repos`](https://github.com/nice-registry/all-the-package-repos/blob/master/scripts/update.js#L51), lets say `unset`, and then handle the special case where `url === null`.

This new feature is quite tricky as there are multiple cases to take into consideration, like a repo that used to have an url but now is null, or the other way around.

File: scripts/update.js (hunk 3 above)
Comment (eridal) [reply scripts/update.js:221]: Please avoid hacks, simply use: 

```js
const url = extractUrl(change) || null
````

**Comment by zarianec** [comment]
@eridal thanks for your review and suggestions. But I didn't get this one:
>My suggestion is to make `extractUrl` simply return the string URL

Isn't it already return a string URL and we parse/validate it later? I don't see what do you want to change in it.

**Comment by zarianec** [comment]
It could be easier to implement this if we have tests for it. I already have tests that cover the update process and caching but as I mentioned before it requires some significant changes to the `update.js` script.

The biggest issue I had is the self-invoking anonymous function which makes it almost impossible to test because of `process.exit`

Currently, I use a workaround to prevent the script from immediate execution like:

```javascript
module.exports = update

if (process.env.NODE_ENV !== 'test') {
  /**
   * Main
   */
  update().then(
    () => {
      console.log(metadata)
      process.exit(0)
    },
    (err) => {
      console.error(err)
      process.exit(1)
    }
  )
}
```

The better way IMO is to split it and move all the update logic to a separate file. `update.js` should only call this logic and finish the process.

@eridal I can push my tests here and we will work on edge cases/new requirements as it seems this PR will never get finished without it.

File: scripts/update.js (hunk 1 above)
Comment (zarianec) [reply scripts/update.js]: btw, what packages should be counted as invalid now? If we add `repos.unset` statistics it will have exactly the same numbers as `stats.invalid`

**Comment by zeke** [comment]
> push my tests here and we will work on edge cases/new requirements

👍🏼

**Comment by zarianec** [comment]
@zeke @eridal pushed tests. I see we already have a `lib` folder. I think it's a good idea to split the `update.js` and move it to `lib` completely. Scripts in `scripts` will simply use it and handle the results.

**Comment by eridal** [comment]
@zarianec I owe you an apology. 

You have been putting a good amount of efforts into this feature. It was easy for me to quickly hack a solution, submit that PR, but doing so disregarded your work in the process. I'm sorry about it and I'm gonna take action to fix that.

This is open source, the code I wrote is from the community, and the more people involved the better; so I'd like you to take over this feature and push it to go live. So I'm gonna go ahead and close my PR, please use it as a source of inspiration --or not, you decide :)

Some ideas behind this PR are super good. My only suggestion is try not to change too much the code, but rather make smaller, incremental changes. If changes are not part of this feature, please submit another PR only for that particular change (ie: mocha update) so that this PR will stay minimal while also making progress. Also it'd be nice if you can execute a full-history reprocess, so more people knows how to do so.

Please accept my apologies, I am totally confident you will be able to deliver this feature!

**Comment by zeke** [comment]
Thanks for chiming in, @eridal. 💛 

> My only suggestion is try not to change too much the code, but rather make smaller, incremental changes.

Totally agree.

> I'd like you to take over this feature and push it to go live. 

Yeah let's do it while the fire is still hot. @zarianec is this PR ready to ship?

**Comment by zarianec** [comment]
@eridal thank you for your understanding, I really appreciate it. Some of the changes in your PR are really helpful and I moved it there (like handling invalid packages and better stats in the readme).

>My only suggestion is try not to change too much the code, but rather make smaller, incremental changes.

I didn't want to refactor the tests as part of this PR and planned to open a new one after we finish with this. So I revert it for now and add the metadata tests you created in #39

>Also it'd be nice if you can execute a full-history reprocess, so more people knows how to do so.

I did it already many times and must say that the caching you implemented is amazing - it saved me a LOT of time.

>is this PR ready to ship?

Unfortunately no. I ran the full rebuild of the index and the metadata test suddenly failed:

```
1) metadata repos should match repos:

      AssertionError: expected 1050411 to equal 1050413
      + expected - actual

      -1050411
      +1050413

  2) metadata repos should match urls:

      AssertionError: expected 712688 to equal 712690
      + expected - actual

      -712688
      +712690
```

It took me some time to figure out why 2 packages are missing and the reason is pretty interesting. There is an NPM package called `constructor` and packages store is a generic javascript object so doing this:
https://github.com/nice-registry/all-the-package-repos/blob/5dd54c54258346bbff779134566f7fdc8f4fe3a8/scripts/update.js#L218

Is the same as:

```javascript
let foo = repos['constructor'] // we simply grab the object's constructor function
console.log(foo) // [Function: Object]
```

The better (and safer) way to handle the arbitrary list of key-value pairs in javascript is `Map` so I'm thinking of using it to store the packages list even considering performance overhead.

```javascript
const packages = new Map(fs.existsSync(files.packages) ? Object.entries(require(files.packages)) : [])
```

File: scripts/update.js [hunk 5]
```diff
@@ -427,7 +427,7 @@ const writeChanges = (deferred) => {
   fs.writeFileSync(files.metadata, toJson(metadata))
 
   if (batch.found > 0) {
-    fs.writeFileSync(files.packages, toJson(sort(packages)))
+    fs.writeFileSync(files.packages, toJson(sort(Object.fromEntries(packages))))
```
Comment (zeke) [inline scripts/update.js:430]: So you're using a Map when collecting data then turning it into an Object when finished? That makes sense to me. 👍🏼

Comment (zarianec) [reply scripts/update.js]: Yes, basically this is it. I updated the `sort` function so it will work with the Map as an input instead of an object - no need to convert the map to an object before sorting it.

File: scripts/update.js [hunk 6]
```diff
@@ -147,25 +149,25 @@ const setupBatch = async (db) => {
...
+const toSortedObject = (packagesMap) => {
   const sorted = {}
-  const keys = Object.keys(object).sort()
+  const keys = [...packagesMap.keys()].sort()
 
   for (const key of keys) {
-    sorted[key] = object[key]
-    delete object[key]
+    sorted[key] = packagesMap.get(key)
+    delete packagesMap.delete(key)
```
Comment (eridal) [inline scripts/update.js:170]: there's an extra `delete` keyword here

File: scripts/update.js [hunk 7]
```diff
@@ -215,43 +209,45 @@ const apply = (change) => {
   stats.changes += 1
 
   const name = change.id
-  const curr = packages[name]
+  const curr = packages.get(name)
 
   if (change.deleted) {
-    if (typeof curr === 'string') {
+    if (typeof curr !== 'undefined') {
```
Comment (eridal) [inline scripts/update.js:215]: You can take advantage of the map: `package.has(name)`

**Comment by eridal** [comment]
Hey, this is looking great!

I wonder if it makes sense that this package exposes the `Map` to its consumers, so that they wont have the same issue that you found here.

File: scripts/update.js (hunk 7 above)
Comment (zarianec) [reply scripts/update.js:215]: It looked strange to me to check with `packages.has()` and use the `curr` variable then. And replacing `curr` with `packages.get()` everywhere looked too wordy.

File: scripts/update.js (hunk 6 above)
Comment (zarianec) [reply scripts/update.js]: fixed, thanks

**Comment by zarianec** [comment]
The issue with a `constructor` key happens only when you try to get the value if it wasn't set before. Since the final packages list has it already, it will work fine until the package will get deleted from the NPM 🙂 

So exposing the Map instead of Object makes sense but we should consider memory overhead for Object->Map conversion.

Or we might expose the Object without a prototype (`Object.create(null)`) to keep the interface the same and make migration easier. But it will add performance overhead anyway.

File: lib/stats-tpl.js [hunk 8]
```diff
@@ -13,16 +13,26 @@ module.exports = {
...
     return dedent`
       <!-- stats -->
       Packages | Count | Percentage
-      -------- | ----- | ----------
-      With repository in package.json | ${total} | 100%
+      :------- | -----:| ----------:
+      With repository | ${notNull} | ${perc(notNull)}%
+      Null repository | ${areNull} | ${perc(areNull)}%
+      **Total** | ${total} | ${perc(total)}%
+
```
Comment (eridal) [inline lib/stats-tpl.js:28]: Nit: unnecessary whitespace here

File: hostnames.js [hunk 9]
```diff
@@ -3,7 +3,7 @@
 const URL = require('url')
 const countValues = require('count-array-values')
 const urls = Object.values(require('.'))
-const hostnames = urls.map(url => URL.parse(url).hostname.replace(/^www\./i, ''))
+const hostnames = urls.filter(u => u).map(url => URL.parse(url).hostname.replace(/^www\./i, ''))
```
Comment (eridal) [inline hostnames.js:6]: Can we use `.filter(Boolean)`?

File: test/repos.js [hunk 10]
```diff
@@ -17,12 +18,12 @@ describe('repos', () => {
...
-    this.timeout(10 * 1000)
+  it('is always a URL or null', function () {
+    this.timeout(30 * 1000)
     const urls = Object.values(repos)
-    urls.forEach(url => {
-      expect(isUrl(url), `${url}`).to.eq(true)
-    })
+
+    expect(urls.some(url => isUrl(url)), 'should contain packages with repos').to.equal(true)
+    expect(urls.some(url => url === null), 'should contain packages without repos').to.equal(true)
```
Comment (eridal) [inline test/repos.js:26]: This is expecting at least a `null`, which will not be the case if this is merged into current master. I'd say use the test to expect what we know it might be there, either a null or an url..

```js
for (const url of Object.values(repos)) {
  expect(url === null || isUrl(url), url).to.equal(true)
}
```

Also nothing prevents the packages to be all set, and while super low probability there could be no-null at some point :))

**Comment by eridal** [comment]
Leaving aside those small comments, this looks ready to ship. Great work @zarianec!

The `Map` can be delayed as consumers of this package will always read whatever is imported as part of the object, so the prototype chain make things like `constructor` or `hasOwnProperty` safer for them.

```js
const foo = {}
console.log(typeof foo.hasOwnProperty) // function
foo.hasOwnProperty = 'nope'
console.log(typeof foo.hasOwnProperty) // string
```

---

Let's talk about release. This a breaking change for others so we need a major version.

**Comment by zeke** [comment]
🍿 Great to see the progress here!

File: test/repos.js (hunk 10 above)
Comment (zarianec) [reply test/repos.js:26]: >which will not be the case if this is merged into current master

Agree, it will fail if we merge it without updating the data. On the other hand master with these changes and old index should be considered broken so probably it's fine.

>while super low probability there could be no-null at some point

The probability of that is so low that I would rather consider it a bug if our index doesn't have any `null` value. It will prevent the situation when packages without a repository are missing in the index because of a bug in the code.

File: hostnames.js (hunk 9 above)
Comment (zarianec) [reply hostnames.js]: Yes, fixed.

**Comment by zeke** [comment]
Sounds good. Wanna have a zoom call soon to pair on the release?

**Comment by zeke** [comment]
> DM me in Twitter

Followed you, but don't see an option to DM just yet. You DM me. :)

**Comment by zarianec** [comment]
@zeke I just checked if I have all required permissions to release this PR. Looks like I'm good to go so I'm going to release this PR tomorrow (need to rebuild index from scratch to eliminate NPM's private packages leakage)

**Comment by zeke** [comment]
Thanks @zarianec

Sorry I totally dropped the ball on this. 🙈
